### PR TITLE
Enhance fs package with file locking and comprehensive testing

### PR DESCRIPTION
**Pull request by amery**
### **User description**

# Enhance fs package with file locking and comprehensive testing

## Summary

This PR enhances the `fs` package with comprehensive file locking capabilities, adds missing test coverage for internal utility functions, and includes documentation and tooling improvements. The changes span the core `fs` package, the `fs/fssyscall` subpackage, and build tooling.

### 🔒 File Locking Enhancements (`fs/fssyscall`)

**Complete TryLockEx Implementation:**

- Add `TryLockEx()` function for non-blocking file lock acquisition.
- **Linux**: Uses `flock()` with `LOCK_NB` flag, converts `EAGAIN`/`EWOULDBLOCK` → `EBUSY`.
- **Windows**: Full implementation using `LockFileEx()` with `LOCKFILE_FAIL_IMMEDIATELY`, converts `ERROR_LOCK_VIOLATION` → `EBUSY`.
- Comprehensive test suite with concurrent locking tests and channel synchronisation.

**Complete Windows File Locking:**

- Replace Windows placeholder implementations with full native API support.
- Use `LockFileEx()` and `UnlockFileEx()` Win32 APIs for all lock operations.
- Lock entire file range with `LOCKFILE_EXCLUSIVE_LOCK` flag.
- Add `golang.org/x/sys/windows` dependency for native Windows API access.

**os.File Wrapper Functions:**

- Add `FLockEx()`, `FUnlockEx()`, and `FTryLockEx()` convenience functions.
- Accept `*os.File` parameters and delegate to Handle-based operations.
- Automatic Handle conversion eliminates manual conversion requirements.
- Nil file validation returns `core.ErrInvalid` for safety.

### 📚 Documentation and Testing (`fs`)

**Comprehensive Documentation:**

- Add detailed file locking section to `fs/README.md` with API documentation.
- Document Handle operations, os.File convenience functions, and cross-platform behaviour.
- Include platform-specific implementation details (Linux flock vs Windows LockFileEx).
- Add Socket.dev security badge for package vulnerability tracking.

**Utility Function Coverage:**

- Add `utils_test.go` with comprehensive test coverage for `joinRunes` and `unsafeCutRoot`.
- Enhanced documentation explaining function behaviour and "unsafe" nature.
- Test coverage includes Unicode paths, edge cases, and error conditions.
- Refactor `unsafeCutRoot` switch from condition-based to value-based matching.

### 🛠️ Build and Tooling Improvements

**Shell Script Quality:**

- Fix shellcheck SC2268 warning in `internal/build/fix_whitespace.sh`.
- Remove unnecessary x-prefix in shell comparison (`"x$last_byte" != "x0a"` → `"$last_byte" != "0a"`).
- Modernise shell scripting patterns for better maintainability.

**Spelling Dictionary:**

- Add `EBUSY` to cspell dictionary for syscall error handling documentation.

## Test Coverage

- **100% coverage** achieved for all `fs/fssyscall` Handle operations including new TryLockEx.
- **100% coverage** for new `os.File` wrapper functions with comprehensive error scenarios.
- **Complete coverage** for `fs` utility functions including Unicode and edge cases.
- **Concurrent testing** with race condition detection and channel synchronisation.
- **Cross-platform compatibility** testing maintained across Linux and Windows.

## Files Changed

- `fs/README.md` ✏️ (+31/-0 lines) - File locking documentation and Socket.dev badge
- `fs/fssyscall/file.go` ➕ (38 lines) - New `os.File` wrapper functions
- `fs/fssyscall/file_test.go` ➕ (150 lines) - Comprehensive wrapper tests
- `fs/fssyscall/syscall_test.go` ➕ (279 lines) - Complete syscall test suite
- `fs/fssyscall/syscall_linux.go` ✏️ (+13/-0 lines) - `TryLockEx` implementation
- `fs/fssyscall/syscall_windows.go` ✏️ (+50/-22 lines) - Complete Windows implementation
- `fs/utils.go` ✏️ (+32/-4 lines) - Enhanced documentation and refactoring
- `fs/utils_test.go` ➕ (121 lines) - Utility function tests
- `fs/go.mod` ✏️ (+3/-1 lines) - Add golang.org/x/sys dependency
- `fs/go.sum` ✏️ (+2/-0 lines) - Dependency checksums
- `internal/build/cspell.json` ✏️ (+1/-0 lines) - Add EBUSY to dictionary
- `internal/build/fix_whitespace.sh` ✏️ (+1/-1 lines) - Fix shellcheck warning

## Benefits

- **Enhanced API**: Convenient `os.File` interface reduces boilerplate code.
- **Non-blocking locks**: `TryLockEx` enables lock acquisition without blocking on both platforms.
- **Full Windows support**: Complete native Windows file locking implementation.
- **Improved maintainability**: Comprehensive test coverage and documentation.
- **Cross-platform consistency**: Identical API and error handling across Linux and Windows.
- **Security awareness**: Socket.dev badge provides vulnerability monitoring.
- **Code quality**: Modern shell scripting and comprehensive spell checking.

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Add file locking wrapper functions for `*os.File`

- Implement non-blocking `TryLockEx` function

- Add comprehensive test coverage for syscall operations

- Enhance documentation for utility functions

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["os.File"] --> B["FLockEx/FUnlockEx/FTryLockEx"]
  B --> C["Handle conversion"]
  C --> D["LockEx/UnlockEx/TryLockEx"]
  D --> E["Platform-specific syscalls"]
  F["Test Suite"] --> G["100% coverage"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>file.go</strong><dd><code>Add os.File wrapper functions for file locking</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

fs/fssyscall/file.go

<ul><li>Add <code>FLockEx</code>, <code>FUnlockEx</code>, and <code>FTryLockEx</code> wrapper functions<br> <li> Accept <code>*os.File</code> parameters and convert to Handle internally<br> <li> Include nil file validation returning <code>core.ErrInvalid</code></ul>

</details>

  </td>
  <td><a href="https://github.com/darvaza-proxy/x/pull/243/files#diff-7be0e7e0bb1a06c42661d2e0e2672a8119a159188d217358e1fce15239fd3777">+38/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>syscall_linux.go</strong><dd><code>Implement non-blocking TryLockEx for Linux</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

fs/fssyscall/syscall_linux.go

<ul><li>Implement <code>TryLockEx</code> function with non-blocking behavior<br> <li> Use <code>LOCK_NB</code> flag and convert <code>EAGAIN</code>/<code>EWOULDBLOCK</code> to <code>EBUSY</code></ul>

</details>

  </td>
  <td><a href="https://github.com/darvaza-proxy/x/pull/243/files#diff-72b0323bca073e972c2ad2f9f7a7042276a09a7f3fc6f9961163e8e285222174">+11/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>syscall_windows.go</strong><dd><code>Add TryLockEx placeholder for Windows</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

fs/fssyscall/syscall_windows.go

<ul><li>Add <code>TryLockEx</code> placeholder function for Windows compatibility<br> <li> Maintain consistent API across platforms with TODO implementation</ul>

</details>

  </td>
  <td><a href="https://github.com/darvaza-proxy/x/pull/243/files#diff-f173394cba9d1687d420ddabbb131f765a6cc7fe9d133eb17b142c44945c0d12">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>file_test.go</strong><dd><code>Add comprehensive tests for file wrapper functions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

fs/fssyscall/file_test.go

<ul><li>Comprehensive test suite for file wrapper functions<br> <li> Test nil file validation and error handling<br> <li> Include lock/unlock sequence and concurrent testing patterns</ul>

</details>

  </td>
  <td><a href="https://github.com/darvaza-proxy/x/pull/243/files#diff-2a0937acf32ee69a2d2c5b09337ce649729b1024340e5c40e10a18a618b23baf">+150/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>syscall_test.go</strong><dd><code>Add comprehensive syscall operation tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

fs/fssyscall/syscall_test.go

<ul><li>Complete test coverage for all Handle operations<br> <li> Concurrent locking tests with channel synchronization<br> <li> Test cases for open, close, lock, unlock, and try-lock operations</ul>

</details>

  </td>
  <td><a href="https://github.com/darvaza-proxy/x/pull/243/files#diff-fe446634064bd526c9ae3d39864ca761fcb7f2a5fe262a5f2b3a02cc29bff9c4">+271/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>utils_test.go</strong><dd><code>Add comprehensive utility function tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

fs/utils_test.go

<ul><li>Complete test coverage for <code>joinRunes</code> and <code>unsafeCutRoot</code> functions<br> <li> Test Unicode paths, edge cases, and error conditions<br> <li> Include exact match, prefix matching, and failure scenarios</ul>

</details>

  </td>
  <td><a href="https://github.com/darvaza-proxy/x/pull/243/files#diff-35641e569127b3a47a068710375ae01ff1dea0cc0234f235b85f881eaf06c024">+121/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>utils.go</strong><dd><code>Enhance utility function documentation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

fs/utils.go

<ul><li>Enhanced documentation for <code>joinRunes</code> and <code>unsafeCutRoot</code> functions<br> <li> Refactor <code>unsafeCutRoot</code> switch from condition-based to value-based<br> <li> Explain "unsafe" nature and usage patterns</ul>

</details>

  </td>
  <td><a href="https://github.com/darvaza-proxy/x/pull/243/files#diff-55c70956b3b0a77b28fe08d078d566654ccbed89ac4a1183dfa2984753c31909">+23/-4</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

## Summary by CodeRabbit

- New Features
  - Added cross-platform advisory exclusive file locking with a non-blocking try-lock that consistently reports "busy" when unavailable.
- Documentation
  - New File Locking usage notes and expanded path-utility explanations and edge cases.
- Refactor
  - Minor internal path-handling control-flow simplification with no behavior change.
- Tests
  - Extensive tests for locking semantics and try-lock behavior; table-driven tests for path utilities.
- Chores
  - Added golang.org/x/sys dependency; small build and spell-check config tweaks.

### CODE REVIEW

File: fs/fssyscall/syscall_test.go [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+		_ = tempFile.Close()
+		_ = os.Remove(tempFile.Name())
+	}()
+
+	tryLockErr := runConcurrentLockTest(tempFile.Name())
+
+	core.AssertError(t, tryLockErr, "try lock on locked file")
+	core.AssertErrorIs(t, tryLockErr, syscall.EBUSY, "error type")
+}
+
```
[the remark this answers is not in the document]
Comment (amery) [reply fs/fssyscall/syscall_test.go:219]: I added windows support for the functions

**Comment by amery** [comment]
Released as v0.5.2: `go get darvaza.org/x/fs@v0.5.2`
